IPython/core/interactiveshell.py

```diff
@@ -2780,9 +2984,14 @@ def error_before_exec(value):
                 self.displayhook.exec_result = result
 
                 # Execute the user code
-                interactivity = 'none' if silent else self.ast_node_interactivity
-                has_raised = self.run_ast_nodes(code_ast.body, cell_name,
-                   interactivity=interactivity, compiler=compiler, result=result)
+                interactivity = "none" if silent else self.ast_node_interactivity
+                if _run_async:
+                    interactivity = 'async'
+                # yield interactivity so let run_cell decide whether to use
+                # an async loop_runner
+                yield interactivity
```

Comment:
Adding `yield interactivity` here makes this a not-entirely-valid coroutine, doesn't it? When I try to call `await run_cell_async` from ipykernel, I get a `bad yield: 'last_expr'`. How can we communicate this required info without faking that this isn't actually a coroutine? Does it need to be an actual async generator?